### ISSUE TITLE
How to use it?

### ISSUE THREAD

**Issue by DRElll** [station384/HekiliHelper#3]
I want to use it, can you tell me how to use it?

**Comment by station384** [comment]
I’ve been working on some documentation. I will get it posted once I settle down on the core.  I keep changing how things work which would invalidate most of the docs I would do right now. On Dec 27, 2023, at 11:06 PM, DRElll ***@***.***> wrote:﻿
I want to use it, can you tell me how to use it?

**Comment by station384** [comment]
Assuming you have compiled this.  Here is the basic setup.
First hekiliHelper will only send F1,F2,F3,F4,F5,F6,F7,F7,F9,F10,F11,F12
It can handle both CTRL and ALT of those keys, excluding ALT-F4 as that would close WOW.
**If you don't want to remap those keys for hekiliHelper purposes, this will not work for you.**
Also in wow CTRL-S and CTRL-TAB should be unmapped in the keybindings.  **If you don't do this you WILL have issues while playing.** 

The basic setup for a single command reading is this.
Copy the CustomerFontLoader  directory contained in this repo to the _retail_\Interface\AddOns This installs a few custom fonts.

In Wow open the hekili options select Multiple, then on the general tab change the icon width to 100, the height to greater than 50.
On the Keybind tab, change the font to Mitra
The font color will need to be different for most classes, you will need to find a color that is not used in any of the icon sets for that class.  for Hunter I use a teal color,  for evoker I use a deep purple.   I have been building some custom profiles for hekili that can be imported.
Change the font size to 40
Position the Hekili Primary command window somewhere out of the way but not directly on the edge of the screen.

HekiliHelper setup.
Load HekiliHelper
Check Stay on top (this is for setup, this can be uncheck after everything is working)
Check Push And Release mode
Uncheck 2nd Image Detection
Click the down arrow on the bottom of hekili helper
Select your activation key, the key you select cannot have an action tied to it.  I use 1, and I leave my 1 command in wow blank (There are reasons for this, as we have to pass thru the command key to wow so when vehicle actions or dragon flight those command will respond) I suggest 1, 2 or 3 whichever you more comfortable holding down.

Click Start

Drag the red selection window over the first (hekili) command in wow
Resize the red selection to fit the hekili command ( not just the text but the command button)
click start (the images should go from black to white)
Click the pick button, the color box will turn black for not color select.
Click on the the keybind text in the game. 
if you click the text the first images in hekili should now be show the hotkey for the button it thinks is next.
You may have to try a few times to hit the color of the font.

Adjust the variance to as low as possible that the system can read.
If it can read it, there will be text under the image showing what hekilihelper thinks it says.
Screen Cap MS is the time between capturing the image on the screen.   4 times a second would be 250.
Key Send MS is the time between command presses when Push And Release mode is NOT checked.

Next for the fun bit.   
You will need to activate 2 command bars,  if your using Vanilla Wow Interface that would be bar 7 and 8, if your using elvUI it would be bar 14 and 15 This is where we are going to put EVERY command Helli may request.
Bar 7 should have the key bindings of F1 - F12 for each button
Bar 8 should have the key bindings of CTRL-F1 - CTRL-F12
When placing the command anything instant cast or with a low cooldown I put on Bar 7 anything rarely used or with a large cooldown 1 minute+ put on bar 8.

Commands that are in Bar 7 or Bar 8  CAN NOT be in one of your 1-9-= hotkey buttons. Hekili will always pick the lowest hotkey if more that on exists, so no duping on the main hotkeys.

(Yeah alot of rules around this, and no flexibility,  working with the limitations of Hekili, and WoW)

once you have some commands on the 7 Bar,  move to a target dummy in wow and press the activation key.
HekiliHelper will start issuing the command is decoded, and wait for the next command and use it when ready. 
If a command appears in Hekili that stalls the attack check and make sure the font color is not in the icon (remember it has to be unique) the command also may not be on bar 7 or 8, check for that.

Any on use equipment your Char has equipped will also need to be on bar 7 or 8, as well as pet summons if you are a hunter or warlock.

This is the basic setup, with this, HekiliHelper will execute the command Hekili suggests,  just not as good as it is capable  doing as in this mode it will not queue up wow commands,  that is where the 2nd Image Detection comes in, and for that we need to change around the hekili windows a lot more, so they are vertical and the queue is the same size as the primary. and also increase the spacing for the queue.  once that is done, you would check the 2nd Image Detection in Hekili Helper,  and blue selection box will appear and this would move over the 2nd hekili command size to fit the button. if its right the 2nd command should decode and you will see the text under the 2nd image.

This is the basic setup for using this.   
Notes:
If you have been playing wow for awhile and have tried may difference builds some weird hekili bugs can show up.  
One of the most frustrating is when Hekili suggest a main command hotkey 1-9 and NOTHING is in the slot.  
There really is something there, but just on a hidden bar that shares the main hotkey.   Wow will use different hidden bars when you change stances or even some talents.   you will need to find them by activating all your bars and changing stances (druid and shadow priest is really effected by this) until you find the hidden command and remove it. Its a pain but not much else can be done.

In theory this command
`/run for i = 1,120 do PickupAction(i) PutItemInBackpack() ClearCursor() end;`
will remove ALL the Actions from your bars even the hidden ones and give you clean slate to work with but I have not tried it yet.

**Comment by DRElll** [comment]
Thank you very much for your explanation.
There are many limitations to using this. I thought it was to detect the current skill icon listed by Hekili through the image and then send the keys corresponding to this skill in the form of a table. For example, you can store the skill pictures displayed on Hekili in an image file, name the picture with the name of the skill, and then list the shortcut keys corresponding to the skill name in a json file. Then automatically send shortcut keys based on the recognized pictures. I feel that this form gives me more freedom to use it, and I don’t need to change my shortcut key settings.

**Comment by station384** [comment]
After writing everything below,  this isn't just a response to you but also notes on decisions I've made and this is as good of place as any to put them for future reference.

TL;DR
I thought of that, doesn't work

-----
For recognizing the pictures Image matching pixel per pixel is easy, 
But that fails due to everyone having different settings,  variations on hue.  graphics quality, anti aliasing etc.  there is no perfect match.  

To expand on this say an image stored on the drive is just solid red.
The game engine will read this, and do some processing on it.   internally it is still red with an RGB value of 255,0,0.

The engine now hands it off the the renderer,
Is the screen its rendering to 8 or 10-bit.  if its 8, don't change anything,  if its 10 change the value of red to what ever the screen is calibrated to so the user will see RED.
It filters it thru anti-aliasing.   this changes the image so now the edges of it at 254,X,Y X and Y being an average of the pixels next to it.
Next it goes thru Hue filters,  the user upped their Hue a bit because they have some color blindness.  now the value is 222,X,Y.
Now when we try and pixel match it will fail.  as 222,X,Y does not equal 255,0,0.

So to correct for this lets put in a little heuristics, and allow for variation.  
222,X,Y is close so give it a probability of 60%, the next one is 80% the next is 90% never going above 90%  because of rendering changes. 
Take those probabilities and average them, you can come up with a 87%.  BUT many icons look the same in WOW,  talking from a pixel perspective,    if you have the same colors in similar quantities just shifted around you will have the same probability. 
Okay lest split the probability blocks up into groups and assign a probability to them based on there average.  then compare the macro block. that way your doing 4 regions. that will eliminate the oblivious pixel shifts, continue down this path and you never get above 90% and the processing approaches that of a neural net with all the iteration that happening.  
The only way around this is to process the image CONTENT, the information being contained in the image itself.  You can do this by training a neural net.  that ups the difficulty in processing 10-100 fold right now, not to mention having to train the model to handle this. and you need a sample set with if not all, at least most of the possible variations that could possibly occur.

Using AI or even a heuristic algorithm to do a probability of likeness on the image works, if in perfect world but one is way to intensive and the other not accurate enough. 

-
Dog leg
I want to do this 30 times a second.  I've given myself a limit of no more than 5ms (60 FPS is 16ms per frame) of processing time so I don't take up CPU and lower game frame rate to much, even with DX12 Wow is still a single core game at its heart for net, input, and logic. that same core also handles frame sync, and syncing up all the threads, if you clog that core to much the game slows.  This is why you see slowdown when there are a lot of other players around and you look at your GPU load and its only at 30% and you CPU on average is only at 30% until you look at the logical CPU activity and see core0 at 100% and the rest at 20% on average  Its that single core bottleneck that is processing all the info from the other players and syncing everything up that is overloaded. Anything dealing with the screen windows also has a preferred core that it executes on,  overload that using programs outside wow, and wow slows down, because it (wow) is using that same screen core.
-

I've done the above in a few of my prototypes.

What I do is I start with the main image that is captured.  mask it with the key color, dispose all the unused pixels that way I have less data to work with and OCR the resulting text, so it doesn't matter what hotkey the user assigns to it, or what the color variation is,   it will press the corresponding key.  

I have the program reading the text, not just matching pixels,  this allows for flexibility if the user changes video settings or has a 10-bit HDR display or an 8-bit SDR.  
The OCR is using the older and faster system,  and I've limited it to what it can recognize to speed it up.   The OCR package and model was originally designed to work on a 486 system, so it has very little impact (people at the time had to be VERY aware of how much time every instruction was taking, so it had to be tight and fast to be useful) and  by me limiting everything to just the bare minimum if is using this only less intensive functionality it hits my time limit I've set, and provides enough flexibility to be used outside a sterile predefined environment.

I've also had to limit the available hot keys to F1-F12 because to bring the OCR recognition probability up to 99% I need AT LEAST 2 characters.   1 looks to much like a 7, - looks to much like a = to the OCR and on those its only right 30% of the time.  So I've made the decision to eliminate them for now.  I may come back later to see if I can tighten this up.

This project came about because of a comment made to me about how they can't play wow anymore or games in general because their hand shakes and they can't hit the key when there playing.   mouse control they can do fine as there hands rest on the table which dampens the movement.

So even with all the concessions that have to be made to set this up in the game to make this work,  it can allow for them to play the game, and return to what they have loved for almost 20 years.   To them its an acceptable tradeoff. 

If I could reduce the tradeoffs I would,  but WoW doesn't have an API for external accessibility aids, and Hekili is a separate project, or I would just change that to smooth out some of the issues and allow for some more flexibility.  example being add prefix and suffix chars to the Hotkey render in hekili.   doing that you can now use a barcode font and just decode a barcode instead of doing OCR,  there would be no ambiguity between 1 and 7 or - and =.  you could also encode cool downs etc into the barcode, better sill hekili render a QR code with all the info encoded in it,  Hotkey, cooldowns, next reccomended command, etc.

I would never ask for those changes as that type of functionality can only be used for automation, boting etc. People don't read barcodes, so what other purpose would it have.   Hekili is for suggesting what the next command should be, giving that information to a person.  its up to the person to choose to press the button or not.  because the person is involved, that avoids it being called a bot.    

This is where this aid falls into a grey area.   It reacts and sends the key for the person depending on what it read.  this is defined in Blizz's TOS as boting anything that can activate a command automatically.  But the helper requires the press of a button to be held to make it go. it isn't automatic, it isn't making any decisions on what to send next or reacting to game conditions, it is just sending the key it sees.  It is very close in concept to an automatic key-press repeater built into a lot of gaming keyboards and using a macro in the game that does a /castsequence which blizz allows.  
Its a grey area for the helper,   but as for hekili itself, it is not.  it is just suggesting the next command.   if they (hekili maintainers) extend it to allow for anything to promote automation it becomes a bot or a tool to promote botting and the project would be banned or broken,  blizz could very easily break it by doing what they do to a lot of the addons,  don't update game variables once combat start.  if they did that hekili would be useless. 

So I'm stuck with the limitations I have been given.   This is my solution.  Its it the best?  don't know.  doubt it as there are people cleverer than myself.   Is it the best I can make it under the ridiculous constraints I have,  maybe.  I know I can improve it, and maybe get it to be accurate with 1-9,-=, that would eliminate the 2 extra bar requirement and remapping for F1-F12 thing.

But that all being said I'm glad I was able to help one person do what they enjoy, and it was fun figuring all this stuff out.
